pallets/asset/src/lib.rs

```diff
@@ -1586,17 +1702,41 @@ impl<T: Trait> Module<T> {
         Ok(())
     }
 
-    fn _update_checkpoint(ticker: &Ticker, user_did: IdentityId, user_balance: T::Balance) {
+    fn _update_checkpoint(ticker: &Ticker, user_did: IdentityId, balance: T::Balance) {
         if <TotalCheckpoints>::contains_key(ticker) {
             let checkpoint_count = Self::total_checkpoints_of(ticker);
             let ticker_user_did_checkpont = (*ticker, user_did, checkpoint_count);
             if !<CheckpointBalance<T>>::contains_key(&ticker_user_did_checkpont) {
-                <CheckpointBalance<T>>::insert(&ticker_user_did_checkpont, user_balance);
+                <CheckpointBalance<T>>::insert(&ticker_user_did_checkpont, balance);
                 <UserCheckpoints>::mutate(&(*ticker, user_did), |user_checkpoints| {
                     user_checkpoints.push(checkpoint_count);
                 });
             }
         }
+        // Record the scheduled checkpoint if one exists and is due.
+        if <NextCheckpoints>::contains_key(ticker, &user_did) {
+            let record_timestamp = T::UnixTime::now().as_secs().saturated_into::<u64>();
+            let schedule_timestamp = Self::next_checkpoints(ticker, &user_did);
+            if schedule_timestamp < record_timestamp {
+                let total_supply = Self::token_details(ticker).total_supply;
+                // Record the checkpoint.
+                <CheckpointRecords<T>>::mutate(ticker, &user_did, |records| {
+                    records.push(CheckpointRecord {
+                        schedule_timestamp,
+                        record_timestamp,
```

Comment:
I don't think we need to save the `record_timestamp` variable. It doesn't matter at all because all the balances at `record_timestamp` and `schedule_timestamp` are the same. All balance changing transactions trigger this function.

Reply:
I'm in favour of having `record_timestamp` because the actual checkpoint records may lag quite a bit if there are no transactions for some time after a checkpoint is due. In fact, some due checkpoints may not get scheduled or recorded 
at all and all such checkpoints will be skipped until after the first transaction. This is because the next checkpoint is only calculated when a transaction is made.